Bugfix badlayers storage

elpaso opened the pull request:
This is a followup of recent changes to the bad layer dialog, the current
approach is to maintain the bad layers (and their position and status in the layer tree)
until the bad layer handling (which is an iterative process) has finished.

There was an issue with the layer store: the previous implementation was just skipping 
a new layer added if a layer with the same id was already in the store, we are
now checking if the layer validity has changed from invalid to valid, and in that case 
we reset the data source of the existing layer.

I believe this makes sense in general: if client code calls an API to add a layer to the store we need to check if the layer is already in the store but we also need to make sure the data source 
can be reset in case it has changed especially when the layer validity has changed from false to true.

The other option would have been to remove the layer before adding it back with the fixed 
data source, but this was leading to a complete loss of the layer configuration (checked/unchecked status , group etc.).

elpaso committed:
Change data source when re-adding an existing layer to the store

The previous implementation was just skipping a layer if
a layer with the same id was already in the store, we are
now checking if the layer validity has changed from invalid
to valid, and in that case we reset the data source of the
existing layer.

Use case: bad layer handling.

elpaso committed:
Minor style changes: same logic as before

elpaso committed:
Added note in the documentation about the new behavior

luipir reviewed (approved):
any idea to add test in test_qgsprojectbadlayers.py

m-kuhn commented on src/core/qgsmaplayerstore.h:
```diff
@@ -149,6 +149,11 @@ class CORE_EXPORT QgsMapLayerStore : public QObject
      *                      the layers yourself. Not available in Python.
      *
+     *
+     * \note if layer with the same id is already in the store it is not added again,
```
```suggestion
     * \note If a layer with the same id is already in the store it is not added again,
```

m-kuhn commented on src/core/qgsmaplayerstore.h:
```diff
@@ -149,6 +149,11 @@ class CORE_EXPORT QgsMapLayerStore : public QObject
      *
+     *
+     * \note if layer with the same id is already in the store it is not added again,
+     *       but if the validity of the layer has changed from false to true, the
```
```suggestion
     *       but if the validity of the layer has changed from FALSE to TRUE, the
```

m-kuhn committed:
Update src/core/qgsmaplayerstore.h

Co-Authored-By: elpaso <elpaso@itopen.it>

m-kuhn committed:
Update src/core/qgsmaplayerstore.h

Co-Authored-By: elpaso <elpaso@itopen.it>

elpaso commented:
> any idea to add test in test_qgsprojectbadlayers.py

Actually no (because it would require a lot of GUI manipulation), but I was thinking at adding a test case for the lower level functionality in the layer storage.

luipir commented:
what bout unit testing only QgsMapLayerStore::addMapLayers  changes? just as partial test

elpaso commented:
> what bout unit testing only QgsMapLayerStore::addMapLayers changes? just as partial test

That's exactly what I meant with:  "I was thinking at adding a test case for the lower level functionality in the layer storage."

elpaso committed:
Add test case for the layer store when re-adding a fixed layer

elpaso commented:
@m-kuhn @nyalldawson the big question here is: should we remove the validity check at 
https://github.com/qgis/QGIS/pull/9620/files#diff-4910c0c1b5527072d2fe858c826b46c4R78
and always reset the data source if it's changed?

Besides of bad layers I don't really see a use case, but on the other hand it looks the most natural way to handle a layer re-add.

nyalldawson commented on src/core/qgsmaplayerstore.cpp:
```diff
@@ -74,6 +74,11 @@ QList<QgsMapLayer *> QgsMapLayerStore::addMapLayers( const QList<QgsMapLayer *>
+    // If the layer is already in the store but its validity has flipped to TRUE reset data source
+    if ( mMapLayers.contains( myLayer->id() ) && ! mMapLayers[myLayer->id()]->isValid() && myLayer->isValid() && myLayer->dataProvider() )
+    {
+      mMapLayers[myLayer->id()]->setDataSource( myLayer->dataProvider()->dataSourceUri(), myLayer->name(), myLayer->providerType(), QgsDataProvider::ProviderOptions() );
```
This leaks myLayer - not sure how to correctly hacked this though (it may not be safe to delete the layer here)

elpaso replied:
@nyalldawson does my last commit address your issue?

I could not come up with a better solution.

elpaso committed:
Fix map layer memory leak with bad layer handling

Long story: if the store contains a layer with the same ID
it silently fails to add it to the store.

This is the cause of leak in case client code calls
store->addLayers() within the (iterative) process of fixing
bad layers by patching the XML and re-reading it.

The proposed solution is to check in QgsProject::addLayer()
- which creates the layer object - if the newly created layer
was already in the store and delete it at the end of the
layer reading process if it was.

Deleting the layer from within the store->addLayers is
extremely dangerous, because the caller could keep a reference
to the layer and use it.

luipir commented on src/core/qgsproject.cpp:
```diff
@@ -995,6 +1001,14 @@ bool QgsProject::addLayer( const QDomElement &layerElem, QList<QDomNode> &broken
   }
+
+  // It should be safe to delete the layer now if layer was stored, because all the store
+  // had to to was to reset the data source in case the validity changed.
```
typo?

elpaso replied:
```suggestion
  // had to do was to reset the data source in case the validity changed.
```
